File: src/AdventOfCode/Puzzles/Y2022/Day22.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode.Puzzles.Y2022;

/// <summary>
/// A class representing the puzzle for <c>https://adventofcode.com/2022/day/22</c>. This class cannot be inherited.
/// </summary>
[Puzzle(2022, 22, "Monkey Map", RequiresData = true, IsHidden = true)]
public sealed class Day22 : Puzzle
{
    /// <summary>
    /// Gets the final password.
    /// </summary>
    public int FinalPassword { get; private set; }

    /// <summary>
    /// Gets the final password by parsing the specified map.
    /// </summary>
    /// <param name="map">The map to parse.</param>
    /// <returns>
    /// The final password derived from <paramref name="map"/>.
    /// </returns>
    public static int GetFinalPassword(IList<string> map)
    {
        var directions = ParseDirections('f' + map[^1]);

        return directions.Count;

        static List<(int Units, char Direction)> ParseDirections(ReadOnlySpan<char> directions)
        {
            var result = new List<(int Units, char Direction)>();

            for (int i = 1; i < directions.Length; i++)
            {
                if (!char.IsDigit(directions[i]))
                {
                    int units = Parse<int>(directions[1..i]);
                    char direction = directions[0];

                    result.Add((units, direction));

                    directions = directions[(i + 1)..];
                    i = -1;
                }
            }

            return result;
        }
    }

    /// <inheritdoc />
    protected override async Task<PuzzleResult> SolveCoreAsync(string[] args, CancellationToken cancellationToken)
    {
        var values = await ReadResourceAsLinesAsync();

        FinalPassword = GetFinalPassword(values);

        if (Verbose)
        {
            Logger.WriteLine("The final password is {0}.", FinalPassword);
        }

        return PuzzleResult.Create(FinalPassword);
    }
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/2024/Day22.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { Puzzle2024 } from './Puzzle2024';

export class Day22 extends Puzzle2024 {
    solution1: number;
    solution2: number;

    override get solved() {
        return false;
    }

    override get name() {
        return '';
    }

    override get day() {
        return 22;
    }

    static solve(values: string[]): number {
        return -1;
    }

    override solveCore(_: string[]) {
        const values = this.readResourceAsLines();

        this.solution1 = Day22.solve(values);
        this.solution2 = Day22.solve(values);

        console.info(`${this.solution1}`);
        console.info(`${this.solution2}`);

        return this.createResult([this.solution1, this.solution2]);
    }
}
